FWE Layout tutorial v2 (#10896)
flutter/website: src/content/get-started/fwe/layout.md

```dart
Widget build(BuildContext context) {
  return Container(
    padding: EdgeInsets.all(16.0),
    child: BorderedImage(),
  );
}
```

The following figure shows a widget without 
padding on the left, 
and a widget with padding on the right.

<img src='/assets/images/docs/fwe/layout/padding.png' alt="A screenshot of a widget with padding and a screenshot of a widget without padding.">

To create more complex layouts in Flutter, 
you can compose many widgets. 
For example, you can combine `Container` and `Center`:

```dart
Widget build(BuildContext context) {
  return Center(
    Container(
      padding: EdgeInsets.all(16.0),
      child: BorderedImage(),
    ),
  );
}
```

## Layout multiple widgets vertically or horizontally

One of the most common layout patterns is to 
arrange widgets vertically or horizontally. 
You can use a `Row` widget to arrange widgets 
horizontally, 
and a `Column` widget to arrange widgets vertically. 
The first figure on this page used both.

This is the most basic example of using a `Row` widget.

{% render docs/code-and-image.md, 
image:"fwe/layout/row.png", 
caption: "This figure shows a row widget with three children."

> alt: "A screenshot of a row widget with three children"

Author: It's a template that puts the code and the image side by side. The alternative would be to use HTML in the file, but I figured it'd be better to make a template since I'm repeating the process so often.

Widgets can be sized to fit within a 
row or column by using the Expanded widget. 
To fix the previous example where the row of 
images is too wide for its render box, 
wrap each image with an [Expanded][] widget.

{% render docs/code-and-image.md,
image:"fwe/layout/expanded_row.png",
caption: "This figure shows a row widget with three children which are wrapped with Expanded widgets."
alt: "A screenshot of three widgets, which take up exactly the amount of space available on the main axis. All three widgets are equal width."
code:"
```dart
Widget build(BuildContext context) {
  return const Row(
    children: [
      [!Expanded!](
        child: BorderedImage(width: 150, height: 150),
      ),
      [!Expanded!](
        child: BorderedImage(width: 150, height: 150),
      ),
      [!Expanded!](
        child: BorderedImage(width: 150, height: 150),
      ),
    ],
  );
}
```
" %}

The Expanded widget can also dictate how much 
space a widget should take up relative
to its siblings. For example,
perhaps you want a widget to occupy twice 
as much space as its siblings. 

> For this, use the Expanded widget flex property, 

Reviewer: ```suggestion
For this, use the `Expanded` widgets `flex` property, 
```

an integer that determines the flex factor 
for a widget. The default flex factor is 1. 
The following code sets the flex factor of the
middle image to 2:

{% render docs/code-and-image.md,
image:"fwe/layout/flex_2_row.png",

> caption: "This figure shows a row widget with three children which are wrapped with Expanded widgets. The center child has it's flex property set to 2."

Reviewer: ```suggestion
caption: "This figure shows a row widget with three children which are wrapped with `Expanded` widgets. The center child has it's `flex` property set to 2."
```

alt: "A screenshot of three widgets, which take up exactly the amount of space available on the main axis. The widget in the center is twice as wide as the widgets on the left and right."
code:"
```dart
Widget build(BuildContext context) {
  return const Row(
    children: [
      Expanded(
        child: BorderedImage(width: 150, height: 150),
      ),
      Expanded(
        [!flex: 2!],
        child: BorderedImage(width: 150, height: 150),
      ),
      Expanded(
        child: BorderedImage(width: 150, height: 150),
      ),
    ],
  );
}
```
" %}

## DevTools and debugging layout

In certain situations, 
a box's constraint is unbounded, or infinite. 
This means that either the maximum width or the

> maximum height is set to [double.infinity][]. 

Reviewer: ```suggestion
maximum height is set to [`double.infinity`][]. 
```

A box that tries to be as big as possible won't
function usefully when given an 
unbounded constraint and, in debug mode, 
throws an exception.

The most common case where a render box ends up
with an unbounded constraint is within a 

> flex box ([Row][] or [Column][]), 

Reviewer: ```suggestion
flex box ([`Row`][] or [`Column`][]), 
```

and within a scrollable region 

> (such as [ListView][] and other [ScrollView][] subclasses). 

Reviewer: ```suggestion
(such as [`ListView`][] and other [`ScrollView`][] subclasses). 
```

> [ListView][], for example, tries to expand to

Reviewer: ```suggestion
`ListView]`, for example, tries to expand to
```

fit the space available in its cross-direction 
(perhaps it's a vertically-scrolling
block and tries to be as wide as its parent). 

> If you nest a vertically scrolling [ListView][] 

Reviewer: ```suggestion
If you nest a vertically scrolling `ListView`
```

> inside a horizontally scrolling ListView, 

Reviewer: ```suggestion
inside a horizontally scrolling `ListView`,
```

the inner list tries to be as wide as possible, 
which is infinitely wide, since the outer one is
scrollable in that direction.

Perhaps the most common error you'll run into 
while building a Flutter application is due to 
incorrectly using layout widgets, 
and is referred to as the "unbounded constraints" 
error.

If there was only one type error you should be 
prepared to confront when you first start building
Flutter apps, it would be this one.

{% ytEmbed 'jckqXR5CrPI', 'Decoding Flutter: Unbounded height and width' %}

:::note The Widget inspector
Flutter has a robust suite of DevTools that
help you work with any number of aspects of
Flutter development.
The "Widget Inspector" tool is particularly
useful when building and debugging layouts (and working with widgets in general).

[Learn more about the Flutter inspector][].
:::

##  Scrolling widgets

Flutter has many built-in widgets that
automatically scroll and also offers a variety of
widgets that you can customize to
create specific scrolling behavior.
On this page, you'll see how to use the most common widget for
making any page scrollable,
as well as a widget for creating scrollable lists.

### ListView

> ListView is a column-like widget that 

Reviewer: ```suggestion
`ListView` is a column-like widget that 
```

automatically provides scrolling when its 
content is longer than its render box.

> The most basic way to use a ListView is 

Reviewer: ```suggestion
The most basic way to use a `ListView` is 
```

> very similar to using a Column or Row. 

Reviewer: ```suggestion
very similar to using a `Column` or `Row`. 
```

Unlike a column or row, 

> a ListView requires its children to take up 

Reviewer: ```suggestion
a `ListView` requires its children to take up 
```

all the available space on the cross axis, 
as shown in the example below. 

{% render docs/code-and-image.md,
image:"fwe/layout/basic_listview.png",
caption: "This figure shows a ListView widget with three children."
alt: "A screenshot of three widgets laid out vertically. They have expanded to take up all available space on the cross axis."
code:"
```dart
Widget build(BuildContext context) {
  return [!ListView!](
    children: const [
      BorderedImage(),
      BorderedImage(),
      BorderedImage(),
    ],
  );
}
```
" %}

> ListViews are commonly used when you have an 

Reviewer: ```suggestion
`ListView`s are commonly used when you have an 
```

unknown or very large (or infinite) number of list items. 
When this is the case, 
it's best to use the `ListView.builder` constructor. 
The builder constructor only builds the 
children that are currently visible on screen.

In the following example, 

> the ListView is displaying a list of to-do items. 

Reviewer: ```suggestion
the `ListView` is displaying a list of to-do items. 
```

things like screen size or input device.
This is referred to as making an app
_adaptive_ and _responsive_.

One of the most useful widgets in making 

> adaptive layouts is the [LayoutBuilder][] widget. 

Reviewer: ```suggestion
adaptive layouts is the [`LayoutBuilder`][] widget. 
```

> LayoutBuilder is one of many widgets that uses

Reviewer: ```suggestion
`LayoutBuilder` is one of many widgets that uses
```

the "builder" pattern in Flutter.

### The builder pattern

In Flutter, you'll find several widgets that use 
the word "builder" in their names or 
in their constructors. 
The following list is not exhaustive:

* [ListView.builder][]
* [Gridview.builder][]
* [Builder][]
* [LayoutBuilder][]
* [FutureBuilder][]

These different "builders" are useful for solving 
different problems. For example, 

> the ListView.builder constructor is primarily used

Reviewer: ```suggestion
the `ListView.builder` constructor is primarily used
```

to lazily render items in a list, 
while the `Builder` widget is useful for gaining 

> access to the BuildContext in deeply widget code.

Reviewer: ```suggestion
access to the `BuildContext` in deeply widget code.
```

Despite their different use cases, 
these builders are unified by how they work. 
Builder widgets and builder constructors all have 
arguments called 'builder' 
(or something similar, 
like `itemBuilder` in the case of `ListView.builder`), 
and the builder argument always accepts a
callback. 
This callback is a __builder function__. 
Builder functions are callbacks that pass data to
the parent widget, 
and the parent widget uses those arguments to 
build and return the child widget.
Builder functions always pass in at least
one argument–the build context–
and generally at least one other argument.

> For example, the LayoutBuilder widget is used 

Reviewer: ```suggestion
For example, the `LayoutBuilder` widget is used 
```

to create responsive layouts based 
on the size of the viewport. The builder callback

> body is passed [BoxConstraints][] its given

Reviewer: ```suggestion
body is passed the [`BoxConstraints`][] that it receives
```

The first time the itemBuilder callback is called 
when Flutter is building the UI, 
the int passed to the function is 0, 
the second time it's 1, and so on.

This allows you to provide specific configuration 
based on the index. Recall the example above using
the`ListView.builder` constructor:

```dart
final List<ToDo> items = Repository.fetchTodos();

Widget build(BuildContext context) {
  return ListView.builder(
    itemCount: items.length,
    itemBuilder: (context, idx) {
      var item = items[idx];
      return Padding(
        padding: const EdgeInsets.all(8.0),
        child: Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: [
            Text(item.description),
            Text(item.isComplete),
          ],
        ),
      );
    },
  );
}
```

This example code uses the index that's 
passed into the builder to grab the correct 
todo from the list of items, 
and then displays that todo's data in 
the widget that is returned from the builder.

To exemplify this, 
the following example changes the 
background color of every other list item.

{% render docs/code-and-image.md,
image:"fwe/layout/alternating_list_items.png"

> caption:"This figure shows a ListView, in which its children have alternating background colors. The background colors were determined programmatically based on the index of the child within the ListView."

Reviewer: ```suggestion
caption:"This figure shows a `ListView`, in which its children have alternating background colors. The background colors were determined programmatically based on the index of the child within the `ListView`."
```

[`Stack`]: {{site.api}}/flutter/widgets/Stack-class.html
[`Positioned`]: {{site.api}}/flutter/widgets/Positioned-class.html
[`MediaQuery`]: {{site.api}}/flutter/widgets/MediaQuery-class.html

> [Transform]:{{site.api}}/flutter/widgets/Transform-class.html

Reviewer: All these need code font within the brackets.
